# Lab 2: Logistic Regression with Python - Brent Samaha

In [1409]:
import sklearn
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
import pandas as pd
import math

## Question 1. Read in dataset 

In [1410]:
spotify_data = pd.read_csv("dataset-of-00s.csv")

In [1411]:
spotify_data.shape

(5872, 19)

#### a) There are 5872 rows (observations) and 19 columns. 

In [1412]:
spotify_data.columns.isna()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False])

#### b) There are no columns with missing values 

In [1413]:
spotify_data.target.value_counts()

1    2936
0    2936
Name: target, dtype: int64

#### c) There are 2936 observations of "1" and 2936 observations of "0" in the Target column

## Question 2. Split the dataset and then separate into training and validation data frames.

In [1414]:
train_data, val_data = train_test_split(spotify_data, test_size = 0.3, random_state = 13)

In [1415]:
train_data.head()

track                     artist  \
979     Livin' Our Love Song      Jason Michael Carroll   
163               Mr. Carter  Lil Wayne Featuring Jay-Z   
2738  To The End Of Our Days                   Alestorm   
3732     Disconnect The Dots                of Montreal   
5597           Bright Lights            matchbox twenty   

                                       uri  danceability  energy  key  \
979   spotify:track:5NhZkLMccDVhxocmVHDu4G         0.488   0.855    7   
163   spotify:track:2tQM1BUrXp2kqOM6mN11TC         0.486   0.710    7   
2738  spotify:track:6tz7QxDke4yIy7IcbyGOeS         0.176   0.911    3   
3732  spotify:track:3n4huoQz0VuJxM5ijcNYno         0.527   0.861    8   
5597  spotify:track:5l36mPDpqNWrNIRzwJpwYl         0.319   0.732    4   

      loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
979     -3.417     1       0.0399      0.287000          0.000000     0.132   
163     -6.287     1       0.3630      0.044500          0.000000     0.350   
2738    -4.443     1       0.0552      0.000047          0.000007     0.383   
3732    -6.554     0       0.0432      0.006110          0.784000     0.113   
5597    -4.475     1       0.0436      0.013400          0.000000     0.658   

      valence    tempo  duration_ms  time_signature  chorus_hit  sections  \
979     0.799  148.971       231467               4    22.76559        13   
163     0.472  170.860       316267               4    20.01532        15   
2738    0.295   93.032       382773               3    26.09560        14   
3732    0.875  136.989       265573               4    29.70212        14   
5597    0.200   80.195       234427               4    28.60857        11   

      target  
979        1  
163        1  
2738       0  
3732       0  
5597       1

In [1416]:
val_data.head()

track           artist  \
3255                    Dum I Huet   Joddla Med Siv   
2957                        Hunger            Phazm   
4456         Doesn't Mean Anything      Alicia Keys   
19             Apply Some Pressure      Maximo Park   
2446  Rockin' Around The Xmas Tree  Joe Lynn Turner   

                                       uri  danceability  energy  key  \
3255  spotify:track:0d8gRz29f7S0SKAJxToU3q         0.703   0.867    0   
2957  spotify:track:21d94acFerDGV1LjkBuCFe         0.330   0.991    0   
4456  spotify:track:5bpx60gYQoDDJcP7vBygPB         0.712   0.407    4   
19    spotify:track:19503qDaxgCdPL2BhJngij         0.378   0.953    9   
2446  spotify:track:4NRX9gOJTO3S9WwXfnXG7Y         0.304   0.975    4   

      loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
3255    -4.309     1       0.0253      0.246000          0.000000    0.3130   
2957    -3.235     0       0.2610      0.000003          0.057700    0.0662   
4456    -7.104     0       0.0308      0.036000          0.000291    0.0963   
19      -3.632     1       0.1380      0.000719          0.000029    0.0936   
2446    -2.468     1       0.1410      0.011600          0.000002    0.2870   

      valence    tempo  duration_ms  time_signature  chorus_hit  sections  \
3255   0.8200   96.006       286427               4    30.77296        12   
2957   0.3900  160.067       241507               4    43.68242         7   
4456   0.0997  103.970       272613               4    47.99257         9   
19     0.5120  167.808       200347               4    20.58661        12   
2446   0.4790  161.876       201800               4    25.79578         9   

      target  
3255       0  
2957       0  
4456       1  
19         0  
2446       0

## Question 3. Additionaly analysis on training set - curiousity, patterns, etc. 

In [1417]:
target = train_data.groupby("target")
hit = target.get_group(1)
flop = target.get_group(0)
train_data.describe()

danceability       energy          key     loudness         mode  \
count   4110.000000  4110.000000  4110.000000  4110.000000  4110.000000   
mean       0.543375     0.694350     5.313625    -7.471226     0.642336   
std        0.191621     0.239684     3.566588     5.103450     0.479371   
min        0.058800     0.000348     0.000000   -47.327000     0.000000   
25%        0.414000     0.565250     2.000000    -8.351750     0.000000   
50%        0.554000     0.746000     5.000000    -6.044000     1.000000   
75%        0.683000     0.887000     8.000000    -4.569500     1.000000   
max        0.986000     0.999000    11.000000     1.137000     1.000000   

       speechiness  acousticness  instrumentalness     liveness      valence  \
count  4110.000000   4110.000000       4110.000000  4110.000000  4110.000000   
mean      0.092353      0.214587          0.153163     0.194662     0.484141   
std       0.093393      0.298006          0.303215     0.159949     0.254300   
min       0.022700      0.000000          0.000000     0.019300     0.000000   
25%       0.036400      0.004370          0.000000     0.093300     0.282250   
50%       0.052400      0.059850          0.000029     0.130000     0.489000   
75%       0.107000      0.314000          0.051150     0.260000     0.686000   
max       0.950000      0.996000          0.998000     0.985000     0.982000   

             tempo   duration_ms  time_signature   chorus_hit     sections  \
count  4110.000000  4.110000e+03     4110.000000  4110.000000  4110.000000   
mean    121.609038  2.590264e+05        3.911192    40.503466    11.074453   
std      30.114864  1.507189e+05        0.411736    20.489948     6.176625   
min      46.755000  1.592000e+04        0.000000     0.000000     1.000000   
25%      97.080250  2.064668e+05        4.000000    27.286925     9.000000   
50%     120.004000  2.379265e+05        4.000000    35.657120    10.000000   
75%     141.055500  2.798865e+05        4.000000    47.532095    12.000000   
max     213.233000  4.170227e+06        5.000000   262.615400   169.000000   

            target  
count  4110.000000  
mean      0.498297  
std       0.500058  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000

In [1418]:
hit.describe()

danceability       energy          key     loudness         mode  \
count   2048.000000  2048.000000  2048.000000  2048.000000  2048.000000   
mean       0.631009     0.715695     5.314941    -5.677040     0.682617   
std        0.151090     0.165805     3.543564     2.050953     0.465571   
min        0.113000     0.107000     0.000000   -16.147000     0.000000   
25%        0.524000     0.604750     2.000000    -6.791750     0.000000   
50%        0.635000     0.736000     5.000000    -5.417500     1.000000   
75%        0.739000     0.849000     8.000000    -4.233000     1.000000   
max        0.986000     0.996000    11.000000    -0.949000     1.000000   

       speechiness  acousticness  instrumentalness     liveness      valence  \
count  2048.000000   2048.000000       2048.000000  2048.000000  2048.000000   
mean      0.098718      0.146582          0.009447     0.185202     0.558306   
std       0.100291      0.181450          0.070303     0.146373     0.222100   
min       0.022700      0.000002          0.000000     0.019300     0.038800   
25%       0.033800      0.014900          0.000000     0.089575     0.378750   
50%       0.049900      0.070100          0.000000     0.130000     0.573000   
75%       0.123250      0.216000          0.000027     0.246000     0.731000   
max       0.628000      0.994000          0.982000     0.985000     0.972000   

             tempo    duration_ms  time_signature   chorus_hit     sections  \
count  2048.000000    2048.000000     2048.000000  2048.000000  2048.000000   
mean    120.396440  238967.587402        3.969238    38.487469    10.478027   
std      29.386939   40453.405499        0.258761    18.115325     2.392487   
min      59.972000  116707.000000        1.000000    13.552980     3.000000   
25%      96.900250  211943.250000        4.000000    26.548443     9.000000   
50%     118.385500  235067.000000        4.000000    34.586585    10.000000   
75%     139.984750  259496.500000        4.000000    44.765320    12.000000   
max     208.078000  561600.000000        5.000000   262.615400    22.000000   

       target  
count  2048.0  
mean      1.0  
std       0.0  
min       1.0  
25%       1.0  
50%       1.0  
75%       1.0  
max       1.0

In [1419]:
flop.describe()

danceability       energy          key     loudness         mode  \
count   2062.000000  2062.000000  2062.000000  2062.000000  2062.000000   
mean       0.456336     0.673151     5.312318    -9.253231     0.602328   
std        0.187941     0.293816     3.590168     6.432138     0.489536   
min        0.058800     0.000348     0.000000   -47.327000     0.000000   
25%        0.314000     0.475250     2.000000   -11.222250     0.000000   
50%        0.460000     0.760000     5.000000    -7.188000     1.000000   
75%        0.597750     0.926750     9.000000    -5.110250     1.000000   
max        0.968000     0.999000    11.000000     1.137000     1.000000   

       speechiness  acousticness  instrumentalness     liveness      valence  \
count  2062.000000   2062.000000       2062.000000  2062.000000  2062.000000   
mean      0.086031      0.282131          0.295904     0.204057     0.410480   
std       0.085554      0.367688          0.370784     0.171898     0.262757   
min       0.023300      0.000000          0.000000     0.023300     0.000000   
25%       0.038900      0.000781          0.000031     0.096050     0.180250   
50%       0.054200      0.038650          0.032300     0.130000     0.384500   
75%       0.096850      0.599750          0.706000     0.277000     0.611000   
max       0.950000      0.996000          0.998000     0.982000     0.982000   

             tempo   duration_ms  time_signature   chorus_hit     sections  \
count  2062.000000  2.062000e+03     2062.000000  2062.000000  2062.000000   
mean    122.813403  2.789490e+05        3.853540    42.505775    11.666828   
std      30.780698  2.070430e+05        0.514595    22.428601     8.346921   
min      46.755000  1.592000e+04        0.000000     0.000000     1.000000   
25%      98.056500  1.932335e+05        4.000000    28.250702     8.000000   
50%     121.953500  2.442335e+05        4.000000    36.853625    10.000000   
75%     142.496750  3.166535e+05        4.000000    50.221705    13.000000   
max     213.233000  4.170227e+06        5.000000   206.927300   169.000000   

       target  
count  2062.0  
mean      0.0  
std       0.0  
min       0.0  
25%       0.0  
50%       0.0  
75%       0.0  
max       0.0

In [1420]:
hit.mean() - flop.mean()

danceability            0.174672
energy                  0.042544
key                     0.002623
loudness                3.576191
mode                    0.080289
speechiness             0.012687
acousticness           -0.135549
instrumentalness       -0.286457
liveness               -0.018855
valence                 0.147826
tempo                  -2.416963
duration_ms        -39981.422782
time_signature          0.115698
chorus_hit             -4.018306
sections               -1.188801
target                  1.000000
dtype: float64

In [1421]:
train_data.key.describe()

count    4110.000000
mean        5.313625
std         3.566588
min         0.000000
25%         2.000000
50%         5.000000
75%         8.000000
max        11.000000
Name: key, dtype: float64

In [1422]:
hit.corr()

danceability    energy       key  loudness      mode  \
danceability          1.000000 -0.056549  0.019840 -0.012815 -0.169309   
energy               -0.056549  1.000000  0.033561  0.652687 -0.069482   
key                   0.019840  0.033561  1.000000  0.002366 -0.121787   
loudness             -0.012815  0.652687  0.002366  1.000000 -0.066453   
mode                 -0.169309 -0.069482 -0.121787 -0.066453  1.000000   
speechiness           0.276850  0.060043  0.044402  0.015937 -0.152938   
acousticness         -0.080854 -0.510548 -0.017093 -0.365818  0.075243   
instrumentalness     -0.050357 -0.022279  0.001054 -0.124223 -0.023676   
liveness             -0.101791  0.178290 -0.007287  0.086616  0.012036   
valence               0.506261  0.364307  0.013860  0.226502 -0.130008   
tempo                -0.319630  0.144871  0.014793  0.082302  0.058278   
duration_ms          -0.002171 -0.182704  0.032454 -0.135554 -0.080382   
time_signature        0.121648  0.090826 -0.010207  0.055374  0.016240   
chorus_hit            0.001053 -0.036957  0.032110 -0.014250 -0.037800   
sections              0.077550 -0.164715 -0.013099 -0.121316 -0.046174   
target                     NaN       NaN       NaN       NaN       NaN   

                  speechiness  acousticness  instrumentalness  liveness  \
danceability         0.276850     -0.080854         -0.050357 -0.101791   
energy               0.060043     -0.510548         -0.022279  0.178290   
key                  0.044402     -0.017093          0.001054 -0.007287   
loudness             0.015937     -0.365818         -0.124223  0.086616   
mode                -0.152938      0.075243         -0.023676  0.012036   
speechiness          1.000000     -0.055556         -0.059350  0.081588   
acousticness        -0.055556      1.000000          0.009222 -0.088508   
instrumentalness    -0.059350      0.009222          1.000000 -0.022618   
liveness             0.081588     -0.088508         -0.022618  1.000000   
valence              0.231692     -0.172418         -0.058598  0.009246   
tempo               -0.059284     -0.070467         -0.008850  0.042597   
duration_ms          0.116548      0.058543          0.027935 -0.003446   
time_signature       0.057690     -0.064784         -0.010700  0.014098   
chorus_hit          -0.003068      0.001631         -0.005372  0.057292   
sections             0.122356      0.061696         -0.005426 -0.037081   
target                    NaN           NaN               NaN       NaN   

                   valence     tempo  duration_ms  time_signature  chorus_hit  \
danceability      0.506261 -0.319630    -0.002171        0.121648    0.001053   
energy            0.364307  0.144871    -0.182704        0.090826   -0.036957   
key               0.013860  0.014793     0.032454       -0.010207    0.032110   
loudness          0.226502  0.082302    -0.135554        0.055374   -0.014250   
mode             -0.130008  0.058278    -0.080382        0.016240   -0.037800   
speechiness       0.231692 -0.059284     0.116548        0.057690   -0.003068   
acousticness     -0.172418 -0.070467     0.058543       -0.064784    0.001631   
instrumentalness -0.058598 -0.008850     0.027935       -0.010700   -0.005372   
liveness          0.009246  0.042597    -0.003446        0.014098    0.057292   
valence           1.000000 -0.066906    -0.179233        0.098987   -0.050437   
tempo            -0.066906  1.000000    -0.046004       -0.028102   -0.060504   
duration_ms      -0.179233 -0.046004     1.000000       -0.010058    0.102396   
time_signature    0.098987 -0.028102    -0.010058        1.000000   -0.004038   
chorus_hit       -0.050437 -0.060504     0.102396       -0.004038    1.000000   
sections         -0.049097  0.094668     0.573944        0.011928   -0.274244   
target                 NaN       NaN          NaN             NaN         NaN   

                  sections  target  
danceability      0.077550     NaN  
energy           -0.164715     

### To do some analysis, I created additional data frames from the training data set. One df that had all training set instances of hits, and another df that had all training set instances of flops. From there, I looked at their stastical descriptions for the numerical columns where I noticed the differences in different properties like danceability, with hits having a higher average. I also noticed potiential outliers like in the flop df, where a minimum value reached -47 for loudness, vastly different than the average in the flop, hit, and overall training dataframe. 

### Knowing that for the "instrumentalness" column that a value above .5 signals a track is considered instrumental, and that subtracting the mean instrumentalness from hits to flops is negative, hits have a lower average of instrumentalness than flops. Further, as the difference in means of the chorus hit between hits and flops is negative, hits on average have their chorus start earlier. Lastly, I looked at the correlation between the variables in the hit dataframe knowing that an assumption of log regression is that predictor variables are not expected to have multicolinearlity. 

## Question 4. Build a logistic regression model

In [1423]:
#Dropping non numeric variables
train_data = train_data.drop(['track', 'artist', 'uri'], axis=1)
val_data = val_data.drop(['track', 'artist', 'uri'], axis=1)

In [1424]:
#Removing target variable 
x_train = train_data.iloc[:,:-1]
x_val = val_data.iloc[:, :-1]

In [1425]:
x_train.dtypes

danceability        float64
energy              float64
key                   int64
loudness            float64
mode                  int64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
duration_ms           int64
time_signature        int64
chorus_hit          float64
sections              int64
dtype: object

In [1426]:
for field in x_train.columns:
    standard_dev = x_train[field].std()
    mean = x_train[field].mean()
    x_train[field] = (x_train[field] - mean) / standard_dev
    x_val[field] = (x_val[field] - mean) / standard_dev

In [1427]:
log_model = LogisticRegression(random_state=13).fit(x_train, train_data.target)

In [1428]:
log_model.predict_proba(x_train)

array([[3.31725247e-01, 6.68274753e-01],
       [3.21149333e-01, 6.78850667e-01],
       [8.68288318e-01, 1.31711682e-01],
       ...,
       [8.16418340e-02, 9.18358166e-01],
       [9.99445591e-01, 5.54408665e-04],
       [1.73578447e-01, 8.26421553e-01]])

In [1429]:
#Getting the target predictions to test against our train_data.target
y_train_pred = log_model.predict(x_train)
y_train_pred

array([1, 1, 0, ..., 1, 0, 1])

In [1430]:
#Getting target predictions to test against our val_data.target
y_val_pred = log_model.predict(x_val)
y_val_pred

array([1, 0, 1, ..., 0, 0, 1])

### Question 4. a)

In [1431]:
#Accuracy:

train_acc = metrics.accuracy_score(train_data.target, y_train_pred)
val_acc = metrics.accuracy_score(val_data.target, y_val_pred)

print("The training dataset accuracy is ",train_acc, "\n","The validation accuracy is ", val_acc, sep="")

The training dataset accuracy is 0.8218978102189781
The validation accuracy is 0.8286038592508513


In [1432]:
#Precision

train_prec = metrics.precision_score(train_data.target, y_train_pred)
val_prec = metrics.precision_score(val_data.target, y_val_pred)

print("The training dataset precision is ",train_prec, "\n","The validation precision is ", val_prec, sep="")

The training dataset precision is 0.789612676056338
The validation precision is 0.7900990099009901


In [1433]:
#Recall

train_rec = metrics.recall_score(train_data.target, y_train_pred)
val_rec = metrics.recall_score(val_data.target, y_val_pred)

print("The training dataset recall is " , train_rec, "\n","The validation recall is " , val_rec, sep="")

The training dataset recall is 0.8759765625
The validation recall is 0.8986486486486487


In [1434]:
#AUC:

In [1435]:
y_train_predprob = log_model.predict_proba(x_train)
y_val_predprob = log_model.predict_proba(x_val)

In [1436]:
fpr, tpr, thresholds = metrics.roc_curve(train_data.target, y_train_predprob[:,1], pos_label = 1)

In [1437]:
train_auc = metrics.auc(fpr, tpr)

In [1438]:
print("The training dataset auc is " , train_auc)

The training dataset auc is  0.8930698398475387


In [1439]:
fpr, tpr, thresholds = metrics.roc_curve(val_data.target, y_val_predprob[:,1], pos_label = 1)
val_auc = metrics.auc(fpr, tpr)

In [1440]:
print("The validation AUC is " , val_auc)

The validation AUC is  0.9006799276393097


In [1441]:
#Being above .5, the AUC details that the model is better than classifying at random

### Question 4. b) 

In [1442]:
#Mapping variables to coefficients

In [1443]:
{col_name : coef for col_name, coef in zip(x_train.columns, log_model.coef_[0])}

{'danceability': 0.99089763197019,
 'energy': -1.5004220531522312,
 'key': 0.03824325159719389,
 'loudness': 2.227920360444547,
 'mode': 0.23140042708524028,
 'speechiness': 0.013787407424749631,
 'acousticness': -0.6068385839404907,
 'instrumentalness': -1.7141431087156203,
 'liveness': -0.04813140606092885,
 'valence': -0.024618535094914983,
 'tempo': 0.05685000041766305,
 'duration_ms': -0.24068314126542129,
 'time_signature': 0.13211195930332092,
 'chorus_hit': -0.0799863570868194,
 'sections': -0.06348529030341142}

## Question 5. Feature engineering - variable transformation 

In [1444]:
#duration until chorus as per song length

x_train["chorus_hit_duration_interaction"] = x_train.chorus_hit  / x_train.duration_ms
x_val["chorus_hit_duration_interaction"] = x_val.chorus_hit / x_val.duration_ms

#### I chose to create this interaction with the thinking that a hit would more likely have their chorus earlier in a song. My thinking is that hits are generally shorter songs, so if a long song is a hit, maybe it's because its chorus comes earlier in the song. By starting earlier in a long song, a listener could be more inclinced to listen to the whole song, like it more, making it a hit. Maybe! 

In [1445]:
x_train['dance_valence'] = (x_train.danceability * x_train.valence)
x_val['dance_valence'] = (x_val.danceability * x_val.valence)

#### I chose to multiply the values of danceability and valence as they had one of the stronger instances of correlation in the training dataset. 

In [1446]:
x_train.loudness.agg(['skew']).transpose()

skew   -2.702339
Name: loudness, dtype: float64

In [1447]:
x_train["loudness_sqr"] = (x_train["loudness"] * x_train["loudness"])
x_val["loudness_sqr"] = (x_val["loudness"] * x_val["loudness"])

#### I chose to square the value of loudness as I looked to the skewd-ness of the value in the training dataset and saw that it was negatively skewed. I believe that by squaring it, there could be some effect to make the distribution more normal. 

In [1448]:
x_train.drop(columns = ['chorus_hit', 'duration_ms', 'loudness', 'danceability', 'valence',], inplace = True)
x_val.drop(columns = ['chorus_hit', 'duration_ms', 'loudness', 'danceability','valence',], inplace = True)

In [1449]:
x_train.describe()

energy           key          mode   speechiness  acousticness  \
count  4.110000e+03  4.110000e+03  4.110000e+03  4.110000e+03  4.110000e+03   
mean  -9.978771e-16  1.058899e-17  8.001169e-17 -4.589989e-15  1.387833e-15   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -2.895488e+00 -1.489834e+00 -1.339956e+00 -7.458044e-01 -7.200769e-01   
25%   -5.386272e-01 -9.290743e-01 -1.339956e+00 -5.991127e-01 -7.054127e-01   
50%    2.154904e-01 -8.793427e-02  7.461117e-01 -4.277938e-01 -5.192420e-01   
75%    8.037647e-01  7.532058e-01  7.461117e-01  1.568318e-01  3.335932e-01   
max    1.271046e+00  1.594346e+00  7.461117e-01  9.183194e+00  2.622138e+00   

       instrumentalness      liveness         tempo  time_signature  \
count      4.110000e+03  4.110000e+03  4.110000e+03    4.110000e+03   
mean      -6.616227e-16  4.707697e-15 -9.823826e-15   -8.190529e-16   
std        1.000000e+00  1.000000e+00  1.000000e+00    1.000000e+00   
min       -5.051321e-01 -1.096362e+00 -2.485618e+00   -9.499268e+00   
25%       -5.051321e-01 -6.337136e-01 -8.145077e-01    2.156910e-01   
50%       -5.050376e-01 -4.042649e-01 -5.329722e-02    2.156910e-01   
75%       -3.364398e-01  4.084962e-01  6.457430e-01    2.156910e-01   
max        2.786265e+00  4.941202e+00  3.042483e+00    2.644431e+00   

           sections  chorus_hit_duration_interaction  dance_valence  \
count  4.110000e+03                      4110.000000    4110.000000   
mean   8.544125e-17                       -18.839586       0.599319   
std    1.000000e+00                      1076.996913       1.020059   
min   -1.631061e+00                    -68029.150889      -2.228437   
25%   -3.358553e-01                        -1.439759      -0.023834   
50%   -1.739546e-01                         0.370888       0.265648   
75%    1.498468e-01                         2.210434       1.048793   
max    2.556826e+01                      3403.503198       4.481726   

       loudness_sqr  
count  4.110000e+03  
mean   9.997567e-01  
std    3.403490e+00  
min    2.298491e-08  
25%    6.014916e-02  
50%    2.279742e-01  
75%    5.553626e-01  
max    6.098946e+01

In [1450]:
log_model2 = LogisticRegression(random_state=13).fit(x_train, train_data.target)

/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [1451]:
y_val_pred2 = log_model2.predict(x_val)

In [1452]:
#Accuracy:

train_acc = metrics.accuracy_score(train_data.target, y_train_pred)
val_acc = metrics.accuracy_score(val_data.target, y_val_pred)

print("The training dataset accuracy is ",train_acc, "\n","The validation accuracy is ", val_acc, sep="")

The training dataset accuracy is 0.8218978102189781
The validation accuracy is 0.8286038592508513


In [1455]:
#Precision

train_prec = metrics.precision_score(train_data.target, y_train_pred)
val_prec = metrics.precision_score(val_data.target, y_val_pred2)

print("The training dataset precision is ",train_prec, "\n","The validation precision is ", val_prec, sep="")

The training dataset precision is 0.789612676056338
The validation precision is 0.7084078711985689


In [1456]:
#Recall

train_rec = metrics.recall_score(train_data.target, y_train_pred)
val_rec = metrics.recall_score(val_data.target, y_val_pred2)

print("The training dataset recall is " , train_rec, "\n","The validation recall is " , val_rec, sep="")

The training dataset recall is 0.8759765625
The validation recall is 0.8918918918918919


In [1457]:
y_train_predprob = log_model2.predict_proba(x_train)
y_val_predprob = log_model2.predict_proba(x_val)

In [1458]:
fpr, tpr, thresholds = metrics.roc_curve(train_data.target, y_train_predprob[:,1], pos_label = 1)

In [1459]:
train_auc2 = metrics.auc(fpr, tpr)

In [1460]:
print("The training dataset auc is " , train_auc2)

The training dataset auc is  0.8259477439606573


In [1461]:
fpr, tpr, thresholds = metrics.roc_curve(val_data.target, y_val_predprob[:,1], pos_label = 1)
val_auc2 = metrics.auc(fpr, tpr)

In [1462]:
print("The validation AUC is " , val_auc2)

The validation AUC is  0.8305650473127589


#### After doing some feature engineering, my evaluation metrics have mostly stayed the same, with some decreasing. Most uniquely, my AUC has decreased the most, by near .07 in both training and validation. 

## Question 6. Which metric is most important? 

#### I believe that the AUC is the most important metric for this specific dataset. I believe this because the AUC details if a model is better suited to pick a random positive instance than a negative one. In this case, the model is better suited to picking a hit, our positive instance, than being chosen by random. With an AUC score of around .90 (before my unhelpful feature engineering), the model proved it has capability as a classifier. 